### PR TITLE
re #10708: test if GNU sed is used

### LINKED ISSUE #10708: `stdlib`'s  `make depend` needs gnused
The `\u` directive [here](https://github.com/ocaml/ocaml/blob/f44236171d937564571eb145fa13ef422ed8f30a/stdlib/Makefile#L269)  is not recognized by macOS's sed. 

Work around: 

```
brew install gnu-sed
cd stdlib
gsed -i 's/sed -E/gsed -E/g' Makefile
make depend
```

### PR DESCRIPTION
**Pull request by alissa-tung**

### CODE REVIEW

File: stdlib/Makefile [hunk 1]
```diff
@@ -264,6 +264,7 @@ depend:
 	  > .depend.tmp
 	$(CAMLDEP) $(DEPFLAGS) -pp "$(AWK) -f ./remove_module_aliases.awk" \
 	  stdlib.ml stdlib.mli >> .depend.tmp
+	$(strings $(which sed) | grep -q 'GNU') || (echo "the sed in PATH is not GNU sed" && exit 1)
```
Comment (Et7f3) [inline stdlib/Makefile]:
```suggestion
	$$(grep -q 'GNU' "$$(which sed)") || {echo "the sed in PATH is not GNU sed" && exit 1}
```

Prefer `{` `}` over `(` `)`
Avoid `|` and dependencies on `strings` and directly call `grep`
Also you need to escape `$` with `$$` or use `$(shell)`
Also always QUOTE!!

Comment (alissa-tung) [reply stdlib/Makefile]: thx, sorry

File: stdlib/Makefile [hunk 2]
```diff
@@ -264,7 +264,7 @@ depend:
 	  > .depend.tmp
 	$(CAMLDEP) $(DEPFLAGS) -pp "$(AWK) -f ./remove_module_aliases.awk" \
 	  stdlib.ml stdlib.mli >> .depend.tmp
-	$(strings $(which sed) | grep -q 'GNU') || (echo "the sed in PATH is not GNU sed" && exit 1)
+	$$(grep -q 'GNU' "$(which sed)") || {echo "the sed in PATH is not GNU sed" && exit 1}
```
Comment (Et7f3) [inline stdlib/Makefile]:
```suggestion
	$$(grep -q 'GNU' "$$(which sed)") || { echo "the sed in PATH is not GNU sed" && exit 1 }
```

My bad I forgot a $ (feel free to squash)

File: stdlib/Makefile [hunk 3]
```diff
@@ -264,7 +264,7 @@ depend:
 	  > .depend.tmp
 	$(CAMLDEP) $(DEPFLAGS) -pp "$(AWK) -f ./remove_module_aliases.awk" \
 	  stdlib.ml stdlib.mli >> .depend.tmp
-	$$(grep -q 'GNU' "$(which sed)") || {echo "the sed in PATH is not GNU sed" && exit 1}
+	$$(grep -q 'GNU' "$$(which sed)") || { echo "the sed in PATH is not GNU sed" && exit 1 }
```
Comment (Et7f3) [inline stdlib/Makefile]:
```suggestion
	grep -q 'GNU' "$$(which sed)" || { echo 'the sed in PATH is not GNU sed' && exit 1; }
```

Ok I tested this should be the last suggestion 😅

**Comment by dra27** [comment]
On both Unix and Windows it's perfectly reasonable to have a binary which can executed but not read, even before getting to the assumption that the string GNU _has_ to appear verbatim in the binary itself!

This patch has the benefit of actually selecting `gsed` automatically on macOS (and other BSDs) if it's there:

```diff
--- a/stdlib/Makefile
+++ b/stdlib/Makefile
@@ -258,13 +258,19 @@ include .depend

 STDLIB_NAMESPACE_MODULES = $(subst $(SPACE),|,$(STDLIB_PREFIXED_MODULES))

+TEST_SED = \
+  $(if $(filter X,$(shell echo x | $(1) -E -e 's/./\u&/' 2>/dev/null)),$(1))
+GNUISH_SED = \
+  $(if $(call TEST_SED,sed),sed,$(if $(call TEST_SED,gsed),gsed,\
+  $(error GNU sed is needed for make depend)))
+
 .PHONY: depend
 depend:
        $(CAMLDEP) $(DEPFLAGS) $(filter-out stdlib.%,$(wildcard *.mli *.ml)) \
          > .depend.tmp
        $(CAMLDEP) $(DEPFLAGS) -pp "$(AWK) -f ./remove_module_aliases.awk" \
          stdlib.ml stdlib.mli >> .depend.tmp
-       sed -E \
+       $(GNUISH_SED) -E \
        -e 's/^(${STDLIB_NAMESPACE_MODULES})(\.[^i]*)(i?) :/\1\2\3 : \1.ml\3/' \
        -e 's#(^| )(${STDLIB_NAMESPACE_MODULES})[.]#\1stdlib__\u\2.#' \
          .depend.tmp > .depend
```

if neither the `sed` nor `gsed` commands support both `-E` and `\u` then you get:

```
Makefile:271: *** GNU sed is needed for make depend. Stop.
```

**Comment by dra27** [comment]
(originally, I'd have envisaged this detection being in `configure`, but in fact it's fine in `stdlib/Makefile` as we really do only use it once)

**Comment by Et7f3** [comment]
> (originally, I'd have envisaged this detection being in `configure`, but in fact it's fine in `stdlib/Makefile` as we really do only use it once)

variables are always evaluated so we don't do once we do each time we invoke make. Please do in configure or use regular shell construct in Makefile (like this PR) so we don't pay one/two invocation each time.

**Comment by dra27** [comment]
Variables are not always evaluated

**Pull request closed without merging**